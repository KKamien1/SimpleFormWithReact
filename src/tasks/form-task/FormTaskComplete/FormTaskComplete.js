
import React, { Component } from 'react'

import { formTaskValidator } from '../utils/validator'
import { saveUserForm } from './form-api'
import InputField from '../components/InputField'
import CheckBox from '../components/CheckBox'

const [username, lastname, birthday, usertype, inactivityDate] = ['username', 'lastname', 'birthday', 'usertype', 'inactivityDate']
export class FormTaskComplete extends Component {
    constructor(props) {
        super(props);
        this.validate = formTaskValidator(this.props);
        this.saveUserForm = saveUserForm;
        this.state = {
            [username]: {
                name: username,
                value: '',
                required: true,
                touched: false,
                errors: [],
                isValid: false
            },
            [lastname]: {
                name: lastname,
                value: '',
                required: true,
                touched: false,
                errors: [],
                isValid: false
            },
            [birthday]: {
                name: birthday,
                value: '',
                touched: false,
                errors: [],
                isValid: false
            },
            [usertype]: {
                name: usertype,
                value: '',
                touched: false,
                errors: [],
                isValid: false
            },
            [inactivityDate]: {
                name: inactivityDate,
                value: '',
                touched: false,
                errors: [],
                isValid: false,
                required: false,
            }
        }
    }

    componentDidMount() {
        this.validateForm();
        this.initialState = { ...this.state }
    }

    isUserTypeInactive = () => this.state['usertype'].value === 'Inactive' ? true : false;

    onChangeHandler = ({ target }) => {
        const { name, value } = target;
        this.setInactivityDateRequired(name, value);
        const { isValid, errors } = this.validate[name](target);
        this.setState((prevState) => ({ [name]: { ...prevState[name], value, isValid, errors } }))
    }

    onBlurHandler = ({ target: { name } }) => {
        this.setState((prevState) => ({ [name]: { ...prevState[name], touched: true } }))
    }

    onSubmit = (event) => {
        event.preventDefault();
        this.saveUserForm();
        this.resetForm()
    }

    validateForm = () => {
        for (const name in this.state) {
            const { isValid, errors } = this.validate[name](this.state[name]);
            this.setState((prevState) => ({ [name]: { ...prevState[name], errors, isValid } }))
        }
    }

    isFormValid = () => {
        return Object.values(this.state).every(field => field.isValid);
    }

    setInactivityDateRequired = (name, value) => {
        if (name === usertype) {
            const required = value === 'Inactive' ? true : false;
            const { isValid, errors } = this.validate[inactivityDate]({ value: this.state[inactivityDate].value, required });
            this.setState((prevState) => ({ [inactivityDate]: { ...prevState[inactivityDate], errors, isValid, required } }))
        }
    }

    resetForm = (event) => {
        event.preventDefault();
        this.setState({ ...this.initialState });
    }

    render() {
        const { username, lastname, birthday, usertype, inactivityDate } = this.state;
        const disableButton = !this.isFormValid();
        return (
            <>
                <form className='form' onSubmit={this.onSubmit} noValidate>
                    <fieldset className='form__fieldset'>
                        <legend className='form__legend'>Personal Info</legend>

                        <InputField
                            field={username}
                            onChange={this.onChangeHandler}
                            onBlur={this.onBlurHandler}
                        >
                            First Name
                        </InputField>

                        <InputField
                            field={lastname}
                            onChange={this.onChangeHandler}
                            onBlur={this.onBlurHandler}
                        >
                            Last Name
                        </InputField>
                        <InputField
                            field={birthday}
                            onChange={this.onChangeHandler}
                            onBlur={this.onBlurHandler}
                            placeholder={this.props.dateFormat}
                        >
                            Birthday
                        </InputField>

                    </fieldset>

                    <fieldset className='form__fieldset'>
                        <legend className='form__legend'>User Management</legend>

                        <CheckBox
                            field={usertype}
                            options={['Active', 'Inactive']}
                            onChange={this.onChangeHandler}
                            checked={true}
                        >
                            User Type
                        </CheckBox>

                        <InputField
                            field={inactivityDate}
                            onChange={this.onChangeHandler}
                            onBlur={this.onBlurHandler}
                            placeholder={this.props.dateFormat}

                        >
                            User Inactivity Date
                        </InputField>
                    </fieldset>
                    <button type='submit' className='form__btn form__btn--primary' disabled={disableButton}>Save</button>
                    <button className='form__btn' onClick={this.resetForm}>Reset</button>
                </form>
            </>
        );
    }
};
