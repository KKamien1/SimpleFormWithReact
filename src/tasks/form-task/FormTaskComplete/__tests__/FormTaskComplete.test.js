import React from 'react';
import { shallow } from 'enzyme';
import { render, screen } from '@testing-library/react';

import { FormTaskComplete } from '../FormTaskComplete'

describe('FormTaskComplete', () => {

    test('should match snapshot', () => {
        const wrapper = shallow(<FormTaskComplete />);
        expect(wrapper.html()).toMatchSnapshot();
    })
    test('should contain form element', () => {
        const wrapper = shallow(<FormTaskComplete />);
        expect(wrapper.find('form')).toBeTruthy()
    })
    test('should contain 2 fieldsets', () => {
        const wrapper = shallow(<FormTaskComplete />);
        expect(wrapper.find('fieldset').length).toBe(2);
    })

    it('renders legend text', () => {
        render(<FormTaskComplete />);
        expect(screen.getByText('Personal Info')).toBeInTheDocument();
        expect(screen.getByText('Personal Info')).toBeInTheDocument();
    })

});