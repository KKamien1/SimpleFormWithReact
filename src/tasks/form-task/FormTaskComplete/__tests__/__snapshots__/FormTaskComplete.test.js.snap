// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTaskComplete should match snapshot 1`] = `"<form class=\\"form\\"><fieldset><legend>Personal Info</legend><div class=\\"form__input-field\\"><label for=\\"username\\">First Name<span class=\\"form__asterisk\\">*</span><input type=\\"text\\" name=\\"username\\" value=\\"\\" required=\\"\\"/></label></div><div class=\\"form__input-field\\"><label for=\\"lastname\\">Last Name<span class=\\"form__asterisk\\">*</span><input type=\\"text\\" name=\\"lastname\\" value=\\"\\" required=\\"\\"/></label></div><div class=\\"form__input-field\\"><label for=\\"birthday\\">Birthday<input type=\\"text\\" name=\\"birthday\\" value=\\"\\"/></label></div></fieldset><fieldset><legend>User Management</legend><div class=\\"form__input-field\\"><label for=\\"usertype\\">User Type<input type=\\"checkbox\\" name=\\"usertype\\" value=\\"\\"/><span>Active</span></label></div><div class=\\"form__input-field\\"><label for=\\"inactivityDate\\">User Inactivity Date<input type=\\"text\\" name=\\"inactivityDate\\" value=\\"\\"/></label></div></fieldset><button>Save</button></form>"`;
